### ISSUE TITLE
wrong optimistic fetch order after online

### ISSUE THREAD

**Issue by eltonec** [FormidableLabs/urql#2347]
urql version & exchanges: "@urql/core": "^2.4.3",    "@urql/exchange-graphcache": "^4.3.6"

**Steps to reproduce**

offline state, order of two operations:
  1.insert todo one
  2.update todo title
 
everything works fine

after online order of automatic fetching:
  1. update todo title -> Not updated because the data does not exist
  2. insert todo one

**Expected behavior**

insert should be before update

**Comment by kitten** [comment]
This is basically expected behaviour. Your mutations are issued and sent in the right order, as in, they're executed in the correct order in which they were issued.

However, because they're sent immediately your backend is receiving them potentially all at once. What that means is that you're executing them immediately, even if you have some kind of client-driven ID to identify resources that have been created while offline.

The only way to solve this is to write an exchange that executes mutations in series. We currently don't have that implemented in urql itself, since we haven't specced out an exchange that does this kind of control, although we were planning on grouping all of this control flow together with reeexecuting on focus, TTL, and other features like that.

Basically, you'll either have to ensure that these kinds of operations cannot be issued while offline (as in, blocking creation) or ensure that you have an exchange that runs certain mutations in series

**Comment by eltonec** [comment]
I got it, thank you very much for the clarification.

**Comment by thephilippbusch** [comment]
Hi, I know this Issue has been closed for quite a while now, but I am facing the same issue right now where optimistic mutations need to be executed in series when the client comes back online (I do not have the possibility to block these mutations). I have been trying to solve this on my own by writing a custom exchange, as pointed out by @kitten.

However, I'm struggling to find the right approach for that exchange. Do I need to provide my own implementation of the fetchExchange with the modifications applied? Or can I simply alter the stream beforehand and pass it along to the existing fetchExchange that does the rest?

I've been trying to figure out a solution for the latter approach, but I can't wrap my head around if it is even possible, to prepare the stream so that the fetchExchange does not need any modifications.

@eltonec I would be super grateful if you could provide some details/example of how you implemented that custom exchange (if you did so).

Any kind of suggestion or pointers in the right direction would be super helpful as I am not so much akin to the way wonka and urql exchanges work in general (just have been crawling the docs/source code and some trial and error).

Thank you so much in advance!

**Comment by kitten** [comment]
Mind you I haven't tested this, but something like this would likely work:

```ts
import {
  makeSubject,
  pipe,
  merge,
  filter,
  onPush,
} from 'wonka';

import type { Exchange, Operation } from '@urql/core';

export const mutationQueueExchange = (): Exchange => {
  return ({ forward }) => {
    return (operations$) => {
      let activeSerialOperation: number | null = null;
      const serialOperationQueue: Operation[] = [];
      const { source: serializedOperations$, next: dispatchOperation } = makeSubject<Operation>();

      const filteredOperations$ = pipe(
        operations$,
        filter((operation) => {
          if (operation.kind !== 'mutation') {
            return true;
          } else if (activeSerialOperation) {
            serialOperationQueue.push(operation);
            return false;
          } else {
            activeSerialOperation = operation.key;
            return true;
          }
        })
      );

      return pipe(
        merge([filteredOperations$, serializedOperations$]),
        forward,
        onPush((result) => {
          if (activeSerialOperation === result.operation.key) {
            const nextOperation = serialOperationQueue.shift();
            if (nextOperation) {
              activeSerialOperation = nextOperation.key;
              dispatchOperation(nextOperation);
            } else {
              activeSerialOperation = null;
            }
          }
        })
      );
    };
  };
};
```

You'd then place that in front of the `fetchExchange` in your configured `exchanges` array.

Any exchange is just like middleware and just like configurable pieces of control flow, as per: https://urql.dev/goto/docs/advanced/authoring-exchanges/
So, the above is actually quite similar to the `retryExchange` we offer. Similar to that, you're free to filter out operations, hold on to them, then re-insert them into the stream for the next exchange (via `forward`) to pick up.

**Comment by thephilippbusch** [comment]
Thanks for the REALLY fast response!! It works like charm!

I've tested the exchange against most of my cases now and everything worked fine! I'll give an update, if an issue arises, but it seems to be a solid solution.

Thanks again!
